onnxruntime/core/graph/graph.cc

```diff
@@ -3753,56 +3794,99 @@ Status Graph::InlineFunction(Node& node) {
   for (auto output_edge : output_edges) {
     RemoveEdge(node.Index(), output_edge.GetNode().Index(), output_edge.GetSrcArgIndex(), output_edge.GetDstArgIndex());
   }
+
+  // Map of function input outputs to nodes input/outputs
   std::unordered_map<std::string, NodeArg*> remap_input_output;
-  if (node.MutableInputDefs().size() != subgraph.GetInputsIncludingInitializers().size())
-    return Status(ONNXRUNTIME, FAIL, "Node " + node.Name() + "'s number of inputs is different from function body graph's number of input.");
+  // Set of node input output names as these names need to be preserved during inlining
+  std::unordered_set<std::string> func_input_output_names;
+
+  ORT_ENFORCE(node.MutableInputDefs().size() == subgraph.GetInputsIncludingInitializers().size(),
+              "Node " + node.Name() + "'s number of inputs is different from function body graph's number of input.");
+
+  ORT_ENFORCE(node.MutableOutputDefs().size() == subgraph.GetOutputs().size(),
+              "Node ", node.Name(), "'s number of outputs is different from function body graph's number of outputs.");
+
   for (size_t i = 0; i < subgraph.GetInputsIncludingInitializers().size(); ++i) {
     auto* input = subgraph.GetInputsIncludingInitializers()[i];
-    if (input->Name() != node.MutableInputDefs()[i]->Name())
+    if (input->Name() != node.MutableInputDefs()[i]->Name()) {
       remap_input_output[input->Name()] = node.MutableInputDefs()[i];
+    }
+    func_input_output_names.insert(input->Name());
   }
 
-  ORT_ENFORCE(node.MutableOutputDefs().size() == subgraph.GetOutputs().size(),
-              "Node ", node.Name(), "'s number of outputs is different from function body graph's number of outputs.");
   for (size_t i = 0; i < subgraph.GetOutputs().size(); ++i) {
     auto* output = subgraph.GetOutputs()[i];
-    if (output->Name() != node.MutableOutputDefs()[i]->Name())
+    if (output->Name() != node.MutableOutputDefs()[i]->Name()) {
       remap_input_output[output->Name()] = node.MutableOutputDefs()[i];
+    }
+    func_input_output_names.insert(output->Name());
   }
 
+  // create a uniq_identifier to append to every node name and intermidiate input\outputs
+  // to make sure there are no unintended duplicates
+  std::stringstream ss;
+  ss << static_cast<const void*>(&node);
+  auto uniq_identifier = ss.str();
+
```

Comment:
Can we use Graph::GenerateNodeArgName?

Reply:
hmm I should have used it. will need to thoroughly test if I make this change. I am working on adding more tests will try to work this in along with that ... otherwise may push this out in a separate PR.